CSLabs.Api/Controllers/ModuleController.cs

```diff
@@ -121,5 +147,25 @@ public async Task<IActionResult> Upsert([FromBody] Module module)
             await DatabaseContext.Entry(module).Collection(m => m.Labs).LoadAsync();
             return Ok(module);
         }
+        
+        [HttpPost("{id}/tags")]
+        public async Task<IActionResult> AddTagToModule(int id, [FromBody] Tag tag)
+        {
+            var module = DatabaseContext.Modules.Single(m => m.Id == id);
+            if (!GetUser().CanEditModules()) {
+                return Forbid("You are not allowed to edit modules");
+            }
+            if (module.Id != 0 && module.OwnerId != GetUser().Id && !GetUser().IsAdmin()) {
+                return Forbid("You are not allowed to edit this module");
+            }
+            // add tag if it does not exist
+            if (DatabaseContext.Tags.Any(t => t.Id == tag.Id))
+            {
+                DatabaseContext.Tags.Add(tag);
+            }
+            module.ModuleTags.Add(new ModuleTag { TagId = tag.Id, ModuleId = module.Id });
+            await DatabaseContext.SaveChangesAsync();
+            return Ok(module);
```

Comment:
Check permissions as early as possible in a request.
Always check if the resource is found first before interacting with it.
You can link of entities using their instances since their Id will be 0 until `DatabaseContext.SaveChangesAsync` is called.
```suggestion
            if (!GetUser().CanEditModules()) {
                return Forbid("You are not allowed to edit modules");
            }
            var module = await DatabaseContext.Modules.FirstOrDefaultAsync(m => m.Id == id);
            if (module == null)
                return NotFound();
            if (module.OwnerId != GetUser().Id && !GetUser().IsAdmin()) {
                return Forbid("You are not allowed to edit this module");
            }
            // add tag if it does not exist
            if (!DatabaseContext.Tags.Any(t => t.Id == tag.Id))
                DatabaseContext.Tags.Add(tag);
            module.ModuleTags.Add(new ModuleTag { Tag = tag, Module = module });
            await DatabaseContext.SaveChangesAsync();
            return Ok(module);
```